### PR TITLE
Add 3rd party dependency version check

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: src/NuGetPackageVerifier/third-party.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    "Microsoft.VisualStudio.Shell.Interop": [ "7.10.6071" ],
+    "Microsoft.VisualStudio.Shell.Interop.10.0": [ "10.0.30319" ],
+    "Microsoft.VisualStudio.Shell.Interop.11.0": [ "11.0.61030" ],
+    "Microsoft.VisualStudio.Shell.Interop.12.0": [ "12.0.30110" ],
+    "Microsoft.VisualStudio.Shell.Interop.8.0": [ "8.0.50727" ],
+    "Microsoft.VisualStudio.Shell.Interop.9.0": [ "9.0.30729" ],
+    "Microsoft.VisualStudio.Text.Data": [ "15.0.26201" ],
+    "Microsoft.Web.Administration": [ "7.0.0" ],
+    "Microsoft.Win32.Registry": [ "4.3.0" ],
+    "Newtonsoft.Json": [ "8.0.3", "10.0.1" ],
```
Comment (ajaybhargavb) [inline src/NuGetPackageVerifier/third-party.json]: Shouldn't this just allow `10.0.1` and have exclusions in specific repos that needs `8.0.3`?

Comment (pakrym) [inline src/NuGetPackageVerifier/third-party.json]: Good idea!

File: src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                    {
+                        yield return issue;
+                    }
+                }
+            }
+        }
+
+        private PackageVerifierIssue VerifyPackageReference(IPackageMetadata contextMetadata, PackageDependencyGroup dependencySet, PackageDependency package)
+        {
+            if (Config.PrefixWhitelist.Any(prefix => package.Id.StartsWith(prefix)))
```
Comment (Eilon) [inline src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs]: Pass in `StringComparison.Ordinal[IgnoreCase]` to this.

File: src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                if (!Config.PrefixBlacklist.Any(prefix => package.Id.StartsWith(prefix)))
```
Comment (Eilon) [inline src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs]: And this.

File: src/NuGetPackageVerifier/PackageIssueFactory.cs [hunk 4]
```diff
@@ -266,12 +266,20 @@ public static PackageVerifierIssue AssemblyHasNoDocFile(string assemblyPath)
...
-        public static PackageVerifierIssue AssemblyHasWrongJsonNetVersion(string assemblyPath, string targetFramework, string currentVersion, string expectedVersion)
+        public static PackageVerifierIssue PackageHasWrongThirdPartyDependencyVersion(string packageId, string targetFramework, string dependencyPackageId, string currentVersion, string expectedVersion)
         {
             return new PackageVerifierIssue(
-                "WRONG_JSONNET_VERSION",
-                string.Format("{0}; {1}", assemblyPath, targetFramework),
-                string.Format("The assembly '{0}' references the wrong Json.NET version. Current version '{1}'; Expected version '{2}'.", assemblyPath, currentVersion, expectedVersion),
+                "WRONG_THIRDPARTY_DEPENDENCY_VERSION",
+                $"{packageId}; {targetFramework}",
+                $"The package '{packageId}' references the wrong version of `{dependencyPackageId}` package. Current version '{currentVersion}'; Expected version '{expectedVersion}'.",
```
Comment (Eilon) [inline src/NuGetPackageVerifier/PackageIssueFactory.cs]: Actually, one more thing: we need to have a wiki page w/ info on what to do if this fails. Can you create a wiki page and add the link to the two errors here? Then we can review what should go into the wiki page.

File: src/NuGetPackageVerifier/PackageIssueFactory.cs [hunk 5]
```diff
@@ -266,12 +266,20 @@ public static PackageVerifierIssue AssemblyHasNoDocFile(string assemblyPath)
...
+                PackageIssueLevel.Error);
+        }
+        public static PackageVerifierIssue PackageHasUnregisteredThirdPartyDependency(string packageId, string targetFramework, string dependencyPackageId, string version)
```
Comment (JunTaoLuo) [inline src/NuGetPackageVerifier/PackageIssueFactory.cs]: nit: insert new line above

File: src/NuGetPackageVerifier/third-party.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
+﻿{
+  "prefixWhitelist": [
+    "Libuv"
+    "Microsoft.AspNetCore",
```
Comment (JunTaoLuo) [inline src/NuGetPackageVerifier/third-party.json]: Curious why some of these prefixes end with dot and not others `Microsoft.AspNetCore` vs `Microsoft.Extensions.`

File: src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                    }
+                }
+            }
+        }
+
+        private PackageVerifierIssue VerifyPackageReference(IPackageMetadata contextMetadata, PackageDependencyGroup dependencySet, PackageDependency package)
+        {
+            if (Config.PrefixWhitelist.Any(prefix => package.Id.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
+            {
+                if (!Config.PrefixBlacklist.Any(prefix => package.Id.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
```
Comment (JunTaoLuo) [inline src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs]: nit: combine nested if statements.

File: src/NuGetPackageVerifier/PackageIssueFactory.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/NuGetPackageVerifier/PackageIssueFactory.cs]: Could probably show this wiki page some love https://github.com/aspnet/Coherence-Signed/wiki/NuGetPackageVerifier

File: src/NuGetPackageVerifier/third-party.json (hunk 6 above)
Comment (pakrym) [inline src/NuGetPackageVerifier/third-party.json]: We ship `Microsoft.AspNetCore` package

File: src/NuGetPackageVerifier/PackageIssueFactory.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/NuGetPackageVerifier/PackageIssueFactory.cs]: It hasn't been updated in a while. Better to file an issue to do that.

File: src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            if (!Config.Packages.TryGetValue(package.Id, out var expectedVersions))
+            {
+                return PackageIssueFactory.PackageHasUnregisteredThirdPartyDependency(
+                    contextMetadata.Id,
+                    dependencySet.TargetFramework.DotNetFrameworkName,
+                    package.Id,
+                    packageVersion);
+            }
+
+            if (!expectedVersions.Contains(packageVersion))
```
Comment (kichalla) [inline src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs]: ignore case

File: src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            if (Config.PrefixWhitelist.Any(prefix => package.Id.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)) &&
+                !Config.PrefixBlacklist.Any(prefix => package.Id.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
+            {
+                // Package is in whitelist and not in the blacklist
+                return null;
+            }
+
+            var packageVersion = package.VersionRange.MinVersion.ToString();
+            if (!Config.Packages.TryGetValue(package.Id, out var expectedVersions))
```
Comment (kichalla) [inline src/NuGetPackageVerifier/Rules/ThirdPartyPackageVersionsRule.cs]: same here

File: src/NuGetPackageVerifier/third-party.json [hunk 10]
```diff
@@ -0,0 +1,74 @@
+﻿{
+  "prefixWhitelist": [
```
Comment (natemcmaster) [inline src/NuGetPackageVerifier/third-party.json]: prefixInclusionList

File: src/NuGetPackageVerifier/third-party.json [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    "Microsoft.EntityFrameworkCore",
+    "Microsoft.Extensions.",
+    "Microsoft.Net.Http.Headers",
+    "Microsoft.NETCore.App",
+    "Microsoft.Owin.Security.Interop",
+    "Microsoft.VisualStudio.Web.",
+    "NETStandard.Library",
+    "System.",
+  ],
+  "prefixBlacklist":  [
```
Comment (natemcmaster) [inline src/NuGetPackageVerifier/third-party.json]: prefixExclusionList

File: src/NuGetPackageVerifier/third-party.json [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    "System.Ben"
```
Comment (natemcmaster) [inline src/NuGetPackageVerifier/third-party.json]: Haha....but remove.

File: src/NuGetPackageVerifier/third-party.json [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  ],
+
+  "packages": {
+    "Microsoft.ApplicationInsights.AspNetCore": [ "2.0.0" ],
+    "Microsoft.AspNet.Identity.EntityFramework": [ "2.2.1" ],
+    "Microsoft.AspNet.WebApi.Client": [ "5.2.2" ],
+    "Microsoft.Azure.KeyVault": [ "2.0.6" ],
+    "Microsoft.Build.Runtime": [ "15.2.0-preview-000047-02" ],
```
Comment (natemcmaster) [inline src/NuGetPackageVerifier/third-party.json]: Can we use floating versions? `15.2.0-*`? This is likely to change frequently, even daily.

Comment (pakrym) [inline src/NuGetPackageVerifier/third-party.json]: Added floated version support

**Pull request merged**